### PR TITLE
Script for spectrogram extraction using teacher forcing and Glow-TTS inference with MAS.

### PR DESCRIPTION
**Pull request by Edresson**

### CODE REVIEW

File: TTS/tts/models/glow_tts.py
Comment (erogol) [inline TTS/tts/models/glow_tts.py:179]: Can you place glow-tts updates to a separate PR?

File: TTS/bin/extract_tts_spectrograms.py
Comment (erogol) [inline TTS/bin/extract_tts_spectrograms.py:1]: we need to implement a test for this script to make sure that the computed spectrograms are right (I don't know how) but at least the script works as we update the rest of the project.

File: TTS/bin/train_vocoder_gan.py
Comment (erogol) [inline TTS/bin/train_vocoder_gan.py:500]: Why do you need this differentiation?

**Review by erogol — CHANGES_REQUESTED** [review]

File: TTS/bin/extract_tts_spectrograms.py
Comment (erogol) [reply TTS/bin/extract_tts_spectrograms.py:1]: Maybe we can compare the features computed by this script with a simpler script loading the model and computing the output spectrogram for just a simple wav file using the same `config.json`.

File: TTS/bin/train_vocoder_gan.py
Comment (Edresson) [reply TTS/bin/train_vocoder_gan.py:500]: I had to do this so that the parameters were in the same order as the official implementation. Thus, it is possible to continue training a model trained in the official implementation. Otherwise, the optimizer breaks.

File: TTS/tts/models/glow_tts.py
Comment (Edresson) [reply TTS/tts/models/glow_tts.py:179]: How should I proceed given that this script depends on these implementations?
Should I undo this PR, submit a PR with only GlowTTS changes, after waiting for the merge and submit this PR again?

File: TTS/bin/train_vocoder_gan.py
Comment (Edresson) [reply TTS/bin/train_vocoder_gan.py:500]: Yes, it's necessary. Otherwise, the training breaks.

File: TTS/bin/extract_tts_spectrograms.py
Comment (Edresson) [reply TTS/bin/extract_tts_spectrograms.py:185]: Do you mean add in: TTS/tts/utils/synthesis.py  or in: TTS/utils/synthesizer.py?

Apparently, the [synthesis function](https://github.com/coqui-ai/TTS/blob/0ee3eeefb553678d56c49534f3972a426a254649/TTS/tts/utils/synthesis.py#L225) does not work with batch so we would have to do another function that receives the model name, text_input, text_lengths, mel_input, mel_lengths, speaker_ids, speaker_embeddings and just return the spectrogram?

Comment (erogol) [reply TTS/bin/extract_tts_spectrograms.py:185]: I mean just creating a function in `extract_tts_spectrograms.py`

File: TTS/tts/models/glow_tts.py
Comment (Edresson) [reply TTS/tts/models/glow_tts.py:179]: I made it

File: TTS/bin/extract_tts_spectrograms.py
Comment (Edresson) [reply TTS/bin/extract_tts_spectrograms.py:1]: Please check if what I did is good enough

File: TTS/bin/extract_tts_spectrograms.py
Comment (erogol) [inline TTS/bin/extract_tts_spectrograms.py:149]: you can replace these lines with `*_`

File: tests/test_glow_tts.py
Comment (erogol) [inline tests/test_glow_tts.py:177]: the same here with `*_`

File: tests/test_extract_tts_spectrograms.py
Comment (erogol) [inline tests/test_extract_tts_spectrograms.py:1]: I guess I couldn't explain well enough. I was suggesting to test ` TTS/bin/extract_tts_spectrograms.py` directly not the model. So if something fails in ` TTS/bin/extract_tts_spectrograms.py` we should know about it. In this case, it is not possible. 

You have two options. 

One is to create a shell script like the those under  "tests/bash_tests/*"

Or something like `test_synthesize.py` to call the script in python using `run_cli`. 

Makes sense ?

**Review by erogol — CHANGES_REQUESTED** [review]

File: tests/test_extract_tts_spectrograms.py
Comment (Edresson) [reply tests/test_extract_tts_spectrograms.py:1]: So I have to train the models and use the checkpoint to extract the spectrograms. Should I add this test within test_glow-tts_train.sh and test_tacotron_train.sh to avoid training the model more than once during the test? Or is it better to create a separate script and train the TTS models and after extract the spectrograms?

Also, should I remove this test or is it better to remove it?

Comment (erogol) [reply tests/test_extract_tts_spectrograms.py:1]: You can just create dummy models without training and save them. Then load them and use with `extract_tts_spectrograms.py`.  Something like `TTS/tests/test_synthesizer.py`. You don't need to train them. 

Or you can use one of the release models to test the script. Either way is fine. 

In this script currently, you are basically testing glow TTS and the Tacotron model. So I think you can remove this.

File: TTS/bin/extract_tts_spectrograms.py
Comment (Edresson) [reply TTS/bin/extract_tts_spectrograms.py]: I made it :)

File: tests/test_extract_tts_spectrograms.py
Comment (Edresson) [reply tests/test_extract_tts_spectrograms.py:1]: Thanks for the explanation, I was confused as my fork did not have `run_cli` and `TTS/tests/test_synthesize.py `so I was looking at `TTS/tests/test_synthesizer.py` and was not understanding exactly what you wanted. Verify Now :).

**Review by erogol — APPROVED** [review]

**Comment by erogol** [comment]
GREAT PR @Edresson !! THX once again.

**Pull request merged**
